src/userLayout/ExplorePage/ExplorePage.tsx

```diff
@@ -0,0 +1,313 @@
+import { AuthContext } from "../../context/AuthContext";
+import { toast } from "react-toastify";
+import imgLogin from "../../assets/images/login PopUp.jpg";
+
+// Style
+const style = {
+  position: "absolute",
+  top: "50%",
+  left: "50%",
+  transform: "translate(-50%, -50%)",
+  width: 400,
+  bgcolor: "background.paper",
+  boxShadow: 24,
+  p: 4,
+};
+
+// interface
+interface IRoom {
+  roomNumber: string;
+  _id: string;
+  price: string;
+  images: string;
+}
+interface State {
+  range?: any;
+  // range?: [Date, Date];
+}
+
+// Function
+export default function ExplorePage() {
+  const navigate = useNavigate();
+  const [roomsList, setRoomsList] = useState<IRoom[]>([]);
+  const [page, setPage] = React.useState(1);
+  const handleChange = (_e: React.ChangeEvent<unknown>, value: number) => {
+    setPage(value);
+  };
+  const location = useLocation();
+  const state = (location.state as State) || {};
+  const roomDateStart = state.range?.[0]?.$d;
+  const roomDateEnd = state.range?.[1].$d;
+  const startDate = dayjs(roomDateStart).format("YYYY-MM-DD");
+  const endDate = dayjs(roomDateEnd).format("YYYY-MM-DD");
+  //modal
+  const [open, setOpen] = React.useState(false);
+  const handleOpen = () => {
+    setOpen(true);
+  };
+  const handleClose = () => setOpen(false);
+
+  // AuthContext
+  const authContext = useContext(AuthContext);
+  if (!authContext) {
+    return null;
+  }
+  const { loginData } = authContext;
+
+  // goToLogin
+  const goToLogin = () => {
+    localStorage.removeItem("token");
+    navigate("/login");
+  };
+
+  // Function getAllRooms
+  const getAllRooms = async (
+    page: number,
+    startDate?: string,
+    endDate?: string
+  ) => {
+    if (!location.state) {
+      try {
+        const response = await axiosInstance.get(`/portal/rooms/available`, {
+          params: {
+            size: 12,
+            page: page,
+          },
+        });
+
+        setRoomsList(response.data.data.rooms);
+      } catch (error) {
+        console.log(error);
+      }
+    }
+    try {
+      const response = await axiosInstance.get(`/portal/rooms/available`, {
+        params: {
+          size: 12,
+          page: page,
+          startDate: startDate,
+          endDate: endDate,
+        },
+      });
+
+      setRoomsList(response.data.data.rooms);
+    } catch (error) {
+      console.log(error);
+    }
+  };
+
+  // Function addfav
+  const addToFav = async (id: string) => {
+    if (!loginData) {
+      handleOpen();
+    } else {
+      try {
+        const response = await axiosInstanceWithHeaders.post(
+          `/portal/favorite-rooms`,
+          { roomId: id }
+        );
+        toast.success(" Add To Favorites Successfully");
+        console.log(response);
+        navigate("/favorites");
+      } catch (error: any) {
+        toast.error(error.response.data.message);
+      }
+    }
+  };
+
+  // UseEffect
+  if (location?.state) {
+    useEffect(() => {
```

Comment:
One of the rules of hooks, that they shouldn't be in conditions like this one. We can move the condition to the inner of the useEffect and call the functions based on the location.state instead.

![image](https://github.com/RanaFathy123/hotel-app/assets/64327685/d2ce4f77-7f12-4456-a9b9-5656f1cecf2f)